lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexFunction.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.vortex.api;
+
+import java.io.Serializable;
+
+/**
+ * Typed user function.
+ * Implement your functions using this interface.
+ *
+ * @param <TInput> input type
+ * @param <TOutput> output type
+ */
+public interface VortexFunction<TInput extends Serializable, TOutput extends Serializable> extends Serializable {
+  /**
+   * This will be called by a TaskletExecutor.
+   */
+  TOutput call(TInput input) throws Exception;
```

Comment:
My idea is to allow users to throw whatever exceptions they want to throw in their `VortexFunction` implementations, catch them in `VortexWorker` to report to `VortexMaster`, and surface them to the user via `ExecutionException` in VortexFuture#get:

```java
public TOutput get() throws InterruptedException, ExecutionException
```

@bgchun I guess we allow them to throw any exception(unchecked+checked) they want, to be caught by `VortexWorker`

@markusweimer Am I understanding correctly that you are referring to a case where the function(thread A) spawns a new thread(thread B) on its own so when thread A throws, B gets orphaned? 

I guess the possibility of arbitrary resource usage by user code is also related with what @zpqian commented before: restricting the amount of resource a tasklet can use. Addressing this can be another new JIRA issue.